cluster-autoscaler/cloudprovider/azure/azure_scale_set.go

```diff
@@ -511,50 +534,113 @@ func (scaleSet *ScaleSet) Nodes() ([]cloudprovider.Instance, error) {
 	splay := rand.New(rand.NewSource(time.Now().UnixNano())).Intn(scaleSet.instancesRefreshJitter + 1)
 	lastRefresh := time.Now().Add(-time.Second * time.Duration(splay))
 
+	orchestrationMode, err := scaleSet.getOrchestrationMode()
+
+	if err != nil {
+		klog.Errorf("failed to get information for VMSS: %s, error: %v", scaleSet.Name, err)
+		return nil, err
+	}
+
+	klog.V(4).Infof("VMSS: orchestration Mode %s", orchestrationMode)
+
+	if orchestrationMode == compute.Uniform {
+		err := buildScaleSetCache(scaleSet, lastRefresh)
+
+		if err != nil {
+			return nil, err
+		}
+
+	} else if orchestrationMode == compute.Flexible {
+		if scaleSet.manager.config.EnableVmssFlex {
+			err := buildScaleSetCacheForFlex(scaleSet, lastRefresh)
+
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, fmt.Errorf("vmss - %q with Flexible orchestration detected but 'enableVmssFlex' feature flag is turned off", scaleSet.Name)
+		}
+
+	} else {
+		return nil, fmt.Errorf("Failed to determine orchestration mode for vmss %q", scaleSet.Name)
+	}
+
+	klog.V(4).Infof("Nodes: returns")
+	return scaleSet.instanceCache, nil
+}
+
+func buildScaleSetCache(scaleSet *ScaleSet, lastRefresh time.Time) error {
```

Comment:
Wouldn't this (and the other one) be better as a method with *ScaleSet receiver?